app/grandchallenge/components/tasks.py

```diff
@@ -1123,46 +1141,65 @@ def add_file_to_object(
     user_upload_pk,
     object_pk,
     interface_pk,
-    civ_pk=None,
+    linked_task=None,
 ):
     from grandchallenge.components.models import (
         ComponentInterface,
         ComponentInterfaceValue,
     )
 
     user_upload = UserUpload.objects.get(pk=user_upload_pk)
-    object = get_model_instance(
-        pk=object_pk,
-        app_label=app_label,
-        model_name=model_name,
+    model = apps.get_model(app_label=app_label, model_name=model_name)
+    object_queryset = model.objects.filter(pk=object_pk).select_for_update(
+        nowait=True
     )
     interface = ComponentInterface.objects.get(pk=interface_pk)
+
+    try:
+        object = object_queryset.get()
+    except OperationalError as error:
+        raise LockNotAcquiredException from error
```

Comment:
What is the guard to ensure that this task is only executed once?

Reply:
Same as above. All we can do, as far as I can see, is make sure that we don't end up with duplicate civs on the object and so we remove the `current_value` if it exists.

Reply:
I was working on catching exceptions in this task and the `add_image_to_object` task. The pattern we'd want is to catch all exceptions, then update the job object status (with an informative error or with something like "An unknown error occurred") an then reraise the exception (unless it is an expected exception that we know how to handle). This pattern however doesn't work if we want to wrap the entire task in a transaction. The job status would never get updated if we raise the exception. So then either, we are ok with not catching all exceptions, or we don't wrap these tasks in transactions. I'm unsure which of these is the way to go. I tend towards the first, only catch what we know how to handle, update the job accordingly, and let all other exceptions get raised without updating the job object and then we can handle them as they come in.

Reply:
Celery has this feature where you specify a task to be executed on task failure (`link_error`), that could be a solution that would allow wrapping the tasks in a transaction but also mark the job as failed if an unexpected error occurs. I don't see it in use anywhere in the codebase though, so maybe another case of "unreliable celery feature"?

Reply:
If you don't want to raise then you do not need to. You should be able to do `logger.error("foo", exc_info=True)` then it would still be reported in the logs and to Sentry.